[Integ-doc] Be more specific on each contracts (#557)
omisego/plasma-contracts: plasma_framework/docs/integration-docs/integration-doc.md

}
```

# Vaults
Vaults are used to deposit funds and, indirectly, to withdraw funds via the exit game.

## Vault contract

> Each vault is a contract holding the custody of the transferred token to Plasma Framework. Each vault would have its own `vaultId` as an identifier used in the Plasma Framework. One can retrieve the vault's address by calling the `PlasmaFramework` with the `vaultId`:

Reviewer: is is transferred to the plasma framework or plasma network?

> Each vault is a contract holding the custody of the transferred token to Plasma Framework. Each vault would have its own `vaultId` as an identifier used in the Plasma Framework. One can retrieve the vault's address by calling the `PlasmaFramework` with the `vaultId`:

Author: My intention is Plasma Framework, which is more contract focused.
Although in practice, the Plasma network of our is heavily defined by the Plasma framework.

}
```

# Vaults
Vaults are used to deposit funds and, indirectly, to withdraw funds via the exit game.

> A deposit to Plasma Framework means that it is moving fund from Root chain (Ethereum) to the Child chain (or Plasma chain). A withdrawal means the opposite, which is the action to move fund from Child chain back to Root chain. A withdrawal must only be indirectly initiated by a canonical exit.

Reviewer 2: ```suggestion
A deposit to Plasma Framework means moving funds from the Root chain (Ethereum) to the Child chain (or Plasma chain). A withdrawal means the opposite, moving funds from the Child chain back to the Root chain. A withdrawal must only be indirectly initiated by a canonical exit.
```

        address indexed token,
        uint256 amount
    );

    event EthWithdrawn(
        address indexed receiver,
        uint256  amount
    );
```

For all events of the vault, see contract documentation: [ETH vault doc](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/contracts/EthVault.md#contract-members)

### ERC20 Vault
```
    event DepositCreated(
        address indexed depositor,
        uint256 indexed blknum,
        address indexed token,
        uint256 amount
    );

    event Erc20Withdrawn(
        address payable indexed receiver,
        address indexed token,
        uint256 amount
    );
```

For all events of the vault, see contract documentation: [ERC20 vault doc](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/contracts/Erc20Vault.md#erc20vaultsol)

# Exit Game

> Exit Games handle all the actions around exits, challenges, etc. Each transaction type would have one and only one exit game contract to enable the direct exit of transaction within that type.

Reviewer: I wonder... do we clarify what *exit* means? Getting out of Plasma Framework and out to Ethereum

> Exit Games handle all the actions around exits, challenges, etc. Each transaction type would have one and only one exit game contract to enable the direct exit of transaction within that type.

Author: add a bit of description of `deposit`, `withdraw`, `exit`. Can you take a look? @InoMurko 

https://github.com/omisego/plasma-contracts/pull/557/commits/b5a3bfcbe412475c1c715965fdf78427ff4e48d8

> Exit Games handle all the actions around exits, challenges, etc. Each transaction type would have one and only one exit game contract to enable the direct exit of transaction within that type.

Reviewer: works for me!

## Exit Game contracts

> Each exit game is a contract that holds the logic and states of a current exit game status. Each exit game is bound to a `txType` as an identifier used in the Plasma Framework. One can retrieve the contract address by calling the `PlasmaFramework` with the `txType`:

Reviewer 2: ```suggestion
An exit game is a contract that holds the logic and states of a current exit game status. Each exit game is bound to a `txType` and uses it as an identifier in the Plasma Framework. You can retrieve an exit game's contract address from the `PlasmaFramework`:
```

`PlasmaFramework.exitGames(txType)`

### Transaction types

> Followings are the `txType` as in currently deployed exit game contracts:

Reviewer 2: ```suggestion
The following are the currently deployed exit game contracts, mapped to `txType`:
```

- Payment (v1): 1
- Fee: 3

> Transaction type 2 is reserved for Payment v2 while no specific exit game contract being registered to the framework yet. For more detail on why we are reserving a transaction type, please see the "Adding new transaction type" section of the [high level design doc](https://docs.google.com/document/d/1PSxLnMskjqje4MksmW2msSSg-GtZoBSMNYey9nEDvt8/edit#heading=h.yu9haziuwf6b)

Reviewer 2: ```suggestion
Transaction type 2 is reserved for Payment v2, although it does not have a specific exit game contract registered in the framework yet. For more detail on why we are reserving a transaction type, please see the "Adding new transaction type" section of the [high level design doc](https://docs.google.com/document/d/1PSxLnMskjqje4MksmW2msSSg-GtZoBSMNYey9nEDvt8/edit#heading=h.yu9haziuwf6b)
```

### Payment Exit Game (V1)

> This contract is the main exit game contract of the first Plasma Framework iteration. It handles the exit game logic, including both standard exit and in-flight exit.

Reviewer 2: ```suggestion
This contract is the main exit game contract in the first Plasma Framework iteration. It handles the exit game logic for payment transactions, including both standard exit and in-flight exit.
```

> Most of the following doc within Exit Game section is about `PaymentExitGame`.

Reviewer 2: ```suggestion
Most of the following documentation in the Exit Game section refers to the `PaymentExitGame`.
```

### Fee Exit Game

> The fee is designed to be exit indirectly from the fee transaction. Operator would first spend the fee transaction output to a payment transaction and then exit the fee value with normal process of exiting a payment transaction.

Reviewer 2: ```suggestion
The fee is designed to be exit indirectly from the fee transaction. Operator must first spend the fee transaction output to a payment transaction and then exit the fee value using the normal process of exiting a payment transaction.
```

        address indexed token,
        uint256 amount
    );

    event EthWithdrawn(
        address indexed receiver,
        uint256  amount
    );
```

For all events of the vault, see contract documentation: [ETH vault doc](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/contracts/EthVault.md#contract-members)

### ERC20 Vault
```
    event DepositCreated(
        address indexed depositor,
        uint256 indexed blknum,
        address indexed token,
        uint256 amount
    );

    event Erc20Withdrawn(
        address payable indexed receiver,
        address indexed token,
        uint256 amount
    );
```

For all events of the vault, see contract documentation: [ERC20 vault doc](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/contracts/Erc20Vault.md#erc20vaultsol)

# Exit Game
Exit Games handle all the actions around exits, challenges, etc. Each transaction type has one and only one exit game contract that provides the means of exiting funds from a transaction of that type.

> Once a exit is canonical to exit after the challenge period, it can be processed according to the exit priority and withdraw the fund back from Child chain to Root chain.

Reviewer 2: ```suggestion
Once an exit is canonical to exit after the challenge period, it can be processed according to the exit priority and withdraws the funds back to the Root chain from the Child chain.
```

## Exit Game contracts
An exit game is a contract that holds the logic and states of a current exit game status. Each exit game is bound to a `txType` and uses it as an identifier in the Plasma Framework. You can retrieve an exit game's contract address from the `PlasmaFramework`:

`PlasmaFramework.exitGames(txType)`

### Transaction types
The following are the currently deployed exit game contracts, mapped to `txType`:
- Payment (v1): 1
- Fee: 3

Transaction type 2 is reserved for Payment v2, although it does not have a specific exit game contract registered in the framework yet. For more detail on why we are reserving a transaction type, please see the "Adding new transaction type" section of the [high level design doc](https://docs.google.com/document/d/1PSxLnMskjqje4MksmW2msSSg-GtZoBSMNYey9nEDvt8/edit#heading=h.yu9haziuwf6b)

### Payment Exit Game (V1)
This contract is the main exit game contract in the first Plasma Framework iteration. It handles the exit game logic for payment transactions, including both standard exit and in-flight exit.

Most of the following documentation in the Exit Game section refers to the `PaymentExitGame`.

### Fee Exit Game

> The fee is designed to be exit indirectly from the fee transaction. Operator must first spend the fee transaction output to a payment transaction and then exit the fee value using the normal process of exiting a payment transaction.

Reviewer 2: ```suggestion
The fee is designed to be exited indirectly from the fee transaction. Operator must first spend the fee transaction output to a payment transaction and then exit the fee value using the normal process of exiting a payment transaction.
```

])
```

### Delete an in-flight exit

> For in-flight exit, an exit is only enqueued to the PlasmaFramework when the exit is piggybacked. As a consequence, if an in-flight exit end up not being piggybacked, the in-flight exit bond would end up being locked.

Reviewer 2: ```suggestion
For in-flight exit, an exit is only enqueued in the PlasmaFramework when the exit is piggybacked. As a consequence, if an in-flight exit is never piggybacked then the in-flight exit bond would end up being locked.
```

> As a mitigation, we have the `deleteNonPiggybackedInFlightExit` api on the exit game contract. One can call this to retrieve the bond back and clean up the state of in-flight exit. This can only be called after the first phase has been passed and nobody piggybacked the exit.

Reviewer 2: ```suggestion
As a mitigation, we have the `deleteNonPiggybackedInFlightExit` api on the exit game contract. One can call this to release the bond back to its owner and clean up the state of in-flight exit. This can only be called after the first phase has been passed and nobody has piggybacked the exit.
```

For some more detail, see the original issue: [here](https://github.com/omisego/plasma-contracts/issues/440)

```
PaymentExitGame.deleteNonPiggybackedInFlightExit(exitId)
```

### Parameters

#### exitId (uint160)
Exit Id of the in-flight exit.

You can get your `exitId` by `getInFlightExitId` helper function of the `PaymentExitGame`, see doc: [here](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/contracts/PaymentExitGame.md#getinflightexitid)

#### Example:

```
PaymentExitGame.deleteNonPiggybackedInFlightExit(707372774235521271159305957085057710072500938)
```

### Processing an in-flight exit

Once the exit period is over, an exit can be processed to release the funds on the root chain. An end user can perform this action, or the operator can do it for everyone.

> Be aware that in-flight exit is only enqueued to PlasmFramework when there exists an piggyback of the token. In other words, if an in-flight exit end up nobody piggyback with certain token, eg. ETH, then user cannot find the exit inside the priority queue for ETH. Please make sure piggyback step is done before process exit. 

Reviewer 2: ```suggestion
Be aware that in-flight exit is only put on a token's exit queue if a piggyback of that token exists. In other words, if an in-flight exit ends up with no piggybacks of a with certain token, eg. ETH, then user will not find the exit on the priority queue for ETH. Please make sure piggyback step is done before process exit. 
```

To process a in-flight exit: 
1. (Optional) Obtain your `exitId` by `getInFlightExitId` helper function of the `PaymentExitGame`, see doc: [here](https://github.com/omisego/plasma-contracts/blob/master/plasma_framework/docs/contracts/PaymentExitGame.md#getinflightexitid)

```
PaymentExitGame.getInFlightExitId(
  "0xf85801c0f4f3019441777dc7bdcc6b58be1c25eb3df7df52d1bfecbd94000000000000000000000000000000000000000087038d7ea4c68000a00000000000000000000000000000000000000000000000000000000000000000", # RLP-encoded transaction sent when startInFlightExit was called
)
```

2. Process your exit. 

> 

Reviewer 2: ```suggestion

Exits are processed in the order of the priority queue (each vault has its own priority queue). To find out which exit is at the head of a queue, you can call `getNextExit(vaultId, token)`. The return value of this call is binary data about the exit and its priority; the exitId is the 160 least significant bits of this data

```

Reviewer: YES! 🎈

```

> PlasmaFramework.processExits({

Reviewer: I've just added this comment: https://github.com/omisego/plasma-contracts/issues/539#issuecomment-573622573
maybe consider covering that case somehow? cc @kevsul

Reviewer 2: I've added a few comments with that extra info. Is that enough, or do we need more details?

Reviewer: yep!!! thanks!

  uint256 vaultId, 
  address token, 
  uint160 topExitId, 
  uint256 maxExitsToProcess
})
```

### Parameters
This section describes the parameters included in the function called for processing a standard exit.

#### vaultId (uint256)
The vault ID of the vault that stores exiting funds.

Use `1` for Ether, `2` for ERC-20.

#### token (address)
The token type to process.

ETH: `0x0000000000000000000000000000000000000000` 

The contract address for ERC-20 tokens.

#### topExitId (uint160)
The unique priority of the first exit that should be processed. Set to zero to skip the check.

> If you're trying to process only your own exit, set your exitId here.

Reviewer 2: ```suggestion
The function of this parameter is to prevent you from inadvertently processing another exit that has jumped to the head of the queue because it has a higher priority. If your exit is at the head of the queue and you want to make sure that you process _only_ your exit then you should set this parameter to your exitId. If another exit with higher priority jumps to the head of the queue, then the processExits() call will fail and you won't spend the gas to process the other exit. You can then wait until your exit is at the head of the queue before trying again.  
```

#### maxExitsToProcess (uint256)
Defines the maximum number of exits you wish to process. Set to `1` to process only your own exit. 

> 

Reviewer 2: ```suggestion
***Note**: `processExits()` will only process exits that have completed their exit period. You can find out which (if any) exits were processed via the `ExitFinalized` or `ProcessedExitsNum` events *

```

Reviewer: is `processExits()` the correct notation? would `processExits/1` be better? (denoting that processExits that we mean here is a function with 1 argument.

Reviewer 2: Hmm, not sure. I've only ever seen `func/arity` notation in Erlang/Elixir though, I don't think it's that common. How about just removing the `()`?

Reviewer: That’s alright () is cool!
